### PR TITLE
webgpu: Restore the separate im2col + matmul for few workgroups

### PR DESCRIPTION
**Pull request by qjia7**
We have several versions of matmul algorithms. Specially we did many optimizations for the small size output.
To utilize those methods for non-1x1 conv2d, we restore the separate im2col + matmul.
 
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7011)

### CODE REVIEW

**Comment by qjia7** [comment]
bodypix-ResNet50-image-0.25 reduces ~5ms and Coco-SSD-MobileNetV2 reduced ~6ms in TGL.

**Review by Linchenn — APPROVED** [review]
Interesting, even though decoupling logics to two shaders introduces overhead, decoupling is still meaningful. LGTM, thank you!

File: tfjs-backend-webgpu/src/im2col_webgpu.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    ${main('index')} {
+      let coords = getCoordsFromIndex(index);
+      if(index < uniforms.size) {
+        let batch = coords[0];
+        let row = ${row};
+        let col = ${col};
+        let offsetY = (row / uniforms.outWidth) * uniforms.stride[0] - uniforms.pad[0];
+        let d0 = offsetY + uniforms.dilation[0] * (col / uniforms.itemsPerBlockRow);
+        var value = 0.0;
+        if(d0 < uniforms.aShape[${rowDim}] && d0 >= 0) {
```
Comment (xhcao) [inline tfjs-backend-webgpu/src/im2col_webgpu.ts:62]: Are `d0` and `d1` equal with `xRow` and `xCol` in the other kernel programs? Do we need to keep them the same.

File: tfjs-backend-webgpu/src/im2col_webgpu.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    const userCode = `
+    ${main('index')} {
+      let coords = getCoordsFromIndex(index);
+      if(index < uniforms.size) {
+        let batch = coords[0];
+        let row = ${row};
+        let col = ${col};
+        let offsetY = (row / uniforms.outWidth) * uniforms.stride[0] - uniforms.pad[0];
+        let d0 = offsetY + uniforms.dilation[0] * (col / uniforms.itemsPerBlockRow);
```
Comment (xhcao) [inline tfjs-backend-webgpu/src/im2col_webgpu.ts:60]: `itemsPerBlockRow` is confused here, someone may consider it as workgroup's feature. But it actually treats input data as a block, it was virtual block. I think it is better to directly transfer `inChannels`  and `filterWidth` as uniform members.

File: tfjs-backend-webgpu/src/kernels/Conv2D_impl.ts [hunk 3]
```diff
@@ -169,6 +170,108 @@ function conv2dByMatMul({
   return out;
 }
 
+// Implements the im2row algorithm as outlined in "High Performance
```
Comment (xhcao) [inline tfjs-backend-webgpu/src/kernels/Conv2D_impl.ts:173]: Is there a conflict between `im2row` here and `im2Col` below?

File: tfjs-backend-webgpu/src/im2col_webgpu.ts (hunk 2 above)
Comment (qjia7) [reply tfjs-backend-webgpu/src/im2col_webgpu.ts]: I would like to keep this name to keep consistent with webgl. And it also saves once calculation `inChannels` * `filterWidth` in shader.

**Comment by qjia7** [comment]
@xhcao @gyagp Please take a another look. I suppose all comments are resolved except the one that I left comment.

File: tfjs-backend-webgpu/src/im2col_webgpu.ts (hunk 2 above)
Comment (xhcao) [reply tfjs-backend-webgpu/src/im2col_webgpu.ts]: I feel OK if you want to keep the original code here.

**Review by xhcao — APPROVED** [review]
LGTM

File: tfjs-backend-webgpu/src/kernels/Conv2D_impl.ts [hunk 4]
```diff
@@ -207,6 +310,27 @@ export function conv2DImpl({
     });
   }
 
+  const thresholdFlagValue = env().getNumber(
+    'WEBGPU_THRESHOLD_TO_INCREASE_WORKGROUPS_FOR_MATMUL');
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/kernels/Conv2D_impl.ts:314]: This flag is for matmul, why do we reuse it for conv2d?

File: tfjs-backend-webgpu/src/flags_webgpu.ts [hunk 5]
```diff
@@ -72,6 +72,14 @@ ENV.registerFlag('WEBGPU_IMPORT_EXTERNAL_TEXTURE', () => true);
...
 
 /**
- * Default Intel execution units number.
+ * Threshold to increase dispatched workgroups for matmul. If too few workgroups
+ * are dispatched, it means the hardware may be in low occupancy.
+ * -1 means it's not set by the user. A default strategy will be applied.
  */
-ENV.registerFlag('WEBGPU_INTEL_EU_COUNT', () => 96);
+ENV.registerFlag(
+    'WEBGPU_THRESHOLD_TO_INCREASE_WORKGROUPS_FOR_MATMUL', () => -1);
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/flags_webgpu.ts:80]: Default it to 0?

File: tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts [hunk 6]
```diff
@@ -115,11 +115,10 @@ export function batchMatMulImpl({
...
     // need to increase workgroups. And the literal number is an empirical
     // value.
-    const thresholdToIncreaseWorkgroups =
-        backend.adapterInfo.intelGPUGeneration >= 12 &&
-            env().get('WEBGPU_INTEL_EU_COUNT') as number >= 96 ?
-        16 :
-        8;
+    const thresholdFlagValue = env().getNumber(
+      'WEBGPU_THRESHOLD_TO_INCREASE_WORKGROUPS_FOR_MATMUL');
+    const thresholdToIncreaseWorkgroups =  thresholdFlagValue > 1 ?
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts:120]: Can we set "WEBGPU_THRESHOLD_TO_INCREASE_WORKGROUPS_FOR_MATMUL" to 1?

File: tfjs-backend-webgpu/src/kernels/Conv2D_impl.ts (hunk 4 above)
Comment (gyagp) [reply tfjs-backend-webgpu/src/kernels/Conv2D_impl.ts:314]: OK

**Review by gyagp — APPROVED** [review]

**Pull request merged**